rskj-core/src/main/java/org/ethereum/core/BlockFactory.java

```diff
@@ -235,51 +296,99 @@ private BlockHeader createBlockHeader(boolean compressed, boolean sealed, byte[]
+        int expectedSizeWithoutMining = calculateExpectedHeaderSize(blockNumber, compressed, false);
+        int expectedSizeWithMining = calculateExpectedHeaderSize(blockNumber, compressed, true);
 
-        return rlpHeader.size() == expectedSize || rlpHeader.size() == expectedSizeMM;
+        return rlpHeader.size() == expectedSizeWithoutMining ||
+                rlpHeader.size() == expectedSizeWithMining;
     }
 
-    private int getRSKIP351SizeAdjustment(long blockNumber, boolean compressed, int preParallelSizeAdjustment) {
-        if (!activationConfig.isActive(ConsensusRule.RSKIP351, blockNumber)) {
-            return 1; // remove version
+    /**
+     * Calculates the expected RLP header size based on active consensus rules.
+     *
+     * @param blockNumber      The block number (used to check activation rules)
+     * @param compressed       Whether compressed encoding is used
+     * @param withMergedMining Whether to include merged mining fields in the calculation
+     * @return The expected number of RLP elements in the header
+     */
+    private int calculateExpectedHeaderSize(long blockNumber, boolean compressed, boolean withMergedMining) {
+        // Expected size with all the optional fields present.
+        int expectedFullSize = withMergedMining ? MAX_RLP_HEADER_SIZE_WITH_MINING : MAX_RLP_HEADER_SIZE_WITHOUT_MINING;
+        // Subtract 1 for each optional field that is NOT present
+        if (!activationConfig.isActive(ConsensusRule.RSKIPUMM, blockNumber)) {
+            // RSKIPUMM is not active, field ummRoot is not present
+            expectedFullSize -= 1;
         }
-
-        if (compressed) {
-            return 2 - preParallelSizeAdjustment; // remove version and edges if existent
+        if (!activationConfig.isActive(ConsensusRule.RSKIP144, blockNumber)) {
+            // Parallel tx execution is not active, field txExecutionSublistsEdges is not present
+            expectedFullSize -= 1;
         }
+        if (!isBaseEventEnabled(blockNumber)) {
+            // baseEvent is not enabled, field baseEvent is not present
+            expectedFullSize -= 1;
+        }
+        expectedFullSize -= getRSKIP351SizeAdjustmentFromExtensionData(blockNumber, compressed);
+
+        return expectedFullSize;
 
-        return 0;
     }
 
-    private static BigInteger parseBigInteger(byte[] bytes) {
-        return bytes == null ? BigInteger.ZERO : BigIntegers.fromUnsignedByteArray(bytes);
+    private boolean isBlockHeaderCompressionEnabled(long blockNumber) {
+        return activationConfig.isActive(ConsensusRule.RSKIP351, blockNumber);
     }
 
-    private static List<Transaction> parseTxs(RLPList txTransactions) {
-        List<Transaction> parsedTxs = new ArrayList<>();
+    private boolean isBaseEventEnabled(long blockNumber) {
+        return (activationConfig.isActive(ConsensusRule.RSKIP535, blockNumber) &&
```

Comment:
The first condition should be enough if we understand that 535 shouldn't be active without 351 right?